aptos-move/framework/supra-stdlib/sources/enumerable_map.move

```diff
@@ -0,0 +1,276 @@
+/// This module provides an implementation of an enumerable map, a data structure that maintains key-value pairs with
+/// efficient operations for addition, removal, and retrieval.
+/// It allows for enumeration of keys in insertion order, bulk operations, and updates while ensuring data consistency.
+/// The module includes error handling and a suite of test functions for validation.
+module supra_std::enumerable_map {
+    use std::error;
+    use std::vector;
+    use aptos_std::table;
+
+    /// Key is already present in the map
+    const EKEY_ALREADY_ADDED: u64 = 1;
+    /// Key is absent in the map
+    const EKEY_ABSENT: u64 = 2;
+    /// Vector is empty
+    const EVECTOR_EMPTY: u64 = 3;
+
+    /// Enumerable Map to store the key value pairs
+    struct EnumerableMap<K : copy + drop, V : store+drop+copy> has store {
+        /// List of all keys
+        list: vector<K>,
+        /// Key mapped to a tuple containing the (position of key in list and value corresponding to the key)
+        map: table::Table<K, Tuple<V>>,
+    }
+
+    /// Tuple to store the position of key in list and value corresponding to the key
+    struct Tuple<V : store+drop+copy> has store, copy, drop {
+        position: u64,
+        value: V,
+    }
+
+    /// Return type
+    struct KeyValue<K : copy + drop, V : store+drop+copy> has store, copy, drop {
+        key: K,
+        value: V,
+    }
+
+    /// To create an empty enum map
+    public fun new_map<K : copy + drop, V : store+drop+copy>(): EnumerableMap<K, V> {
+        return EnumerableMap<K, V> { list: vector::empty<K>(), map: table::new<K, Tuple<V>>() }
+    }
+
+    /// Add Single Key in the Enumerable Map
+    public fun add_value<K : copy+drop, V : store+drop+copy>(map: &mut EnumerableMap<K, V>, key: K, value: V) {
+        assert!(!contains(map, key), error::already_exists(EKEY_ALREADY_ADDED));
+        table::add(&mut map.map, key, Tuple<V> { position: vector::length(&map.list), value });
+        vector::push_back(&mut map.list, key);
+    }
+
+    /// Add Multiple Keys in the Enumerable Map
+    public fun add_value_bulk<K: copy+drop, V : store+drop+copy>(
```

Comment:
Just like `vector::for_each` and `vector::for_each_ref`, it may be helpful if we add `enumerable_map::for_each_value` and `enumerable_map::for_each_value_ref`

Reply:
will be addressed in scope https://github.com/Entropy-Foundation/smr-moonshot/issues/1425 ticket